server/svix-server/src/webhook.rs

```diff
@@ -0,0 +1,319 @@
+// SPDX-FileCopyrightText: © 2022 Svix Authors
+// SPDX-License-Identifier: MIT
+
+use http::{HeaderMap, HeaderValue};
+use time::OffsetDateTime;
+
+#[derive(thiserror::Error, Debug)]
+pub enum WebhookError {
+    #[error("failed to parse timestamp")]
+    InvalidTimestamp,
+
+    #[error("invalid secret")]
+    InvalidSecret(#[from] base64::DecodeError),
+
+    #[error("invalid header {0}")]
+    InvalidHeader(String),
+
+    #[error("signature timestamp too old")]
+    TimestampTooOldError,
+
+    #[error("signature timestamp too far in future")]
+    FutureTimestampError,
+
+    #[error("missing header {0}")]
+    MissingHeader(String),
+
+    #[error("signature invalid")]
+    InvalidSignature,
+
+    #[error("payload invalid")]
+    InvalidPayload,
+}
+
+pub struct Webhook {
+    key: Vec<u8>,
+}
+
+const PREFIX: &str = "whsec_";
+const SVIX_MSG_ID_KEY: &str = "svix-id";
+const SVIX_MSG_SIGNATURE_KEY: &str = "svix-signature";
+const SVIX_MSG_TIMESTAMP_KEY: &str = "svix-timestamp";
+const UNBRANDED_MSG_ID_KEY: &str = "webhook-id";
+const UNBRANDED_MSG_SIGNATURE_KEY: &str = "webhook-signature";
+const UNBRANDED_MSG_TIMESTAMP_KEY: &str = "webhook-timestamp";
+const ID: &str = "id";
+const SIGNATURE: &str = "signature";
+const TIMESTAMP: &str = "timestamp";
+const TOLERANCE_IN_SECONDS: i64 = 5 * 60;
+const SIGNATURE_VERSION: &str = "v1";
+
+impl Webhook {
+    pub fn new(secret: String) -> Result<Self, WebhookError> {
+        let secret = secret.strip_prefix(PREFIX).unwrap_or(&secret);
+        let key = base64::decode(secret)?;
+
+        Ok(Webhook { key })
+    }
+
+    pub fn verify(&self, payload: &[u8], headers: &HeaderMap) -> Result<(), WebhookError> {
```

Comment:
Verify should return the payload (like in other libs).

Reply:
Will add -- note that it's inconsistent across libs. I was consulting java and go, neither of which return anything.

Reply:
Oh right, we only return it if we parse. OK, don't change it.